optimum/onnxruntime/modeling_diffusion.py

```diff
@@ -465,35 +538,133 @@ def dtype(self):
+
+        self._device = device
+
+        return self
+
+    def _get_output_shapes(self, **model_inputs: torch.Tensor) -> Dict[str, int]:
+        known_symbols = self._known_symbols.copy()
+
+        for input_name, compiled_input_shape in self._compiled_input_shapes.items():
+            input_tensor_shape = model_inputs[input_name].shape
+            for expr, dim in zip(compiled_input_shape, input_tensor_shape):
+                if len(expr.free_symbols) == 0:
+                    continue
+                elif len(expr.free_symbols) == 1:
+                    symbol = expr.free_symbols.pop()
+                    if expr == symbol:
+                        known_symbols[symbol] = dim
+                    else:
+                        resolved_symbol = sp.solve(expr - dim, symbol)
+                        if resolved_symbol:
+                            known_symbols[symbol] = int(resolved_symbol[0])
+                        else:
+                            raise ValueError(
+                                f"Failed to resolve the symbolic dimension {symbol} for the input {input_name}. "
+                                f"Expression: {expr}, Dimension: {dim}"
+                            )
+                else:
+                    raise ValueError(
+                        f"Symbolic dimension for the input {input_name} has more than one free symbol. "
+                        f"Expression: {expr}, Dimension: {dim}"
+                    )
+
+        resolved_output_shapes = {}
+        unresolved_output_shapes = {}
+
+        for output_name, compiled_output_shape in self._compiled_output_shapes.items():
+            shape = tuple(dim.subs(known_symbols) for dim in compiled_output_shape)
+
+            if any(dim.free_symbols for dim in shape):
+                unresolved_output_shapes[output_name] = shape
+            else:
+                resolved_output_shapes[output_name] = shape
+
+        if unresolved_output_shapes:
+            raise ValueError(
+                f"Failed to resolve the symbolic dimensions for the model {self.__class__.__name__}. "
+                f"Unresolved output shapes: {unresolved_output_shapes}"
+            )
+
+        return resolved_output_shapes
+
+    def _prepare_io_binding(self, model_inputs: torch.Tensor) -> Tuple[ort.IOBinding, Dict[str, torch.Tensor]]:
+        io_binding = self.session.io_binding()
+
+        for input_name in self.input_names.keys():
+            input_dtype = TypeHelper.ort_type_to_torch_type(self.input_dtypes[input_name])
+
+            if model_inputs[input_name].dtype != input_dtype:
+                model_inputs[input_name] = model_inputs[input_name].to(input_dtype)
+
+            io_binding.bind_input(
+                name=input_name,
+                device_type=self.device.type,
+                device_id=self.device.index if self.device.index is not None else -1,
+                element_type=TypeHelper.ort_type_to_numpy_type(self.input_dtypes[input_name]),
```

Comment:
For onnxruntime 1.20 or later, recommend using onnx type instead of numpy type here. It is because numpy does not support bfloat16, float8; but onnx type supports it.

The mapping from ort type to onnx type is like:
{
   "tensor(float)":  onnx.TensorProto.FLOAT,
   "tensor(float16)":  onnx.TensorProto.FLOAT16,
   ...
}